content/client-lib-development-guide/features.textile

```diff
@@ -468,7 +468,7 @@ h3(#realtime-presence). Presence
 * @(RTP1)@ When a channel @ATTACHED@ @ProtocolMessage@ is received, the @ProtocolMessage@ may contain a @HAS_PRESENCE@ bit flag indicating that there are currently members present on the channel, see "TR3":#TR3 . If the flag is 1, the server will shortly perform a @SYNC@ operation as described in "RTP18":#RTP18 . If that flag is 0 or there is no @flags@ field, the presence map should be considered in sync immediately with no members present on the channel
 * @(RTP2)@ A @PresenceMap@ should be used to maintain a list of members present on a channel. Broadly, this is is a map of "memberKeys":#TP3h to presence messages, all with @PRESENT@ actions (during a sync there may also be ones with an @ABSENT@ action, see "RTP2f":#RTP2f ). Consult the "the Java implementation":https://github.com/ably/ably-java/blob/master/lib/src/main/java/io/ably/lib/realtime/Presence.java#L578-L723 for details.
 ** @(RTP2a)@ All incoming presence messages must be compared for newness with the matching member already in the @PresenceMap@, if one exists, where "matching" means they share the same @memberKey@ (or equivalently, they share both @connectionId@ and @clientId@)
-** @(RTP2b)@ 'Comparing for newness' involves: If the @id@ of an incoming presence message corresponds to its @connectionId@ (defined as the @connectionId@ being an initial substring of the @id@), then the messages are compared lexicographically by @id@. If not, they are compared by timestamp
+** @(RTP2b)@ 'Comparing for newness' involves: If the @id@ of both presence message corresponds to the @connectionId@ (defined as the @connectionId@ being an initial substring of the @id@), then the messages are compared lexicographically by @id@. If not, they are compared by timestamp
```

Comment:
So, summarising the seven possibilities we have so far:

1. Client library compares msgSerial and index (if comparable, else timestamp). Either:

  1. the client lib parses the msgSerial (and index, for presence messages in `SYNC`s) out of the id (possibly changing `id` format to `aaaaa:0:0` rather than `aaaaa-0:0` so client libs can just split on `:`), or:

  1. realtime starts sending `msgSerial` fields in presence messages in `PRESENCE`s, and both `msgSerial` and `index` fields for presence messages in `SYNC`s. (optionally at this point we can drop the `id` from the wire protocol and have the client lib construct it from the `connectionId`, `msgSerial`, and `index`).

1. Client library compares msgSerial only (if comparable, else timestamp). Either:

  1. Client libs all guarantee that no two `PresenceMessages` with the same `clientId` would be sent in the same `ProtocolMessage`, or:

  1. Realtime guarantees to process `PRESENCE`s atomically, see https://github.com/ably/docs/pull/143#discussion_r73927089 .

1. Client library compares new logical clock field. Either:

  1. Clock is a per-connectionId integer that realtime increments for each new msgSerial and for each index within a msgSerial, or:

  1. Same as previous, but per-channel rather than per-connectionId (so like a channelSerial but per `PresenceMessage` rather than per `ProtocolMessage`), or:

  1. If we guarantee that no more than 999 `presenceMessages` per `protocolMessage`: use `msgSerial` + `index`/1000 (e.g. 34.002); client lib can then compare numerically.